txscript/opcode.go

```diff
@@ -1972,90 +1972,37 @@ func opcodeCheckSig(op *opcode, data []byte, vm *Engine) error {
-	// the data stack.  This is required because the more general script
-	// validation consensus rules do not have the new strict encoding
-	// requirements enabled by the flags.
-	hashType := SigHashType(fullSigBytes[len(fullSigBytes)-1])
-	sigBytes := fullSigBytes[:len(fullSigBytes)-1]
-	if err := vm.checkHashTypeEncoding(hashType); err != nil {
-		return err
-	}
-	if err := vm.checkSignatureEncoding(sigBytes); err != nil {
-		return err
-	}
-	if err := vm.checkPubKeyEncoding(pkBytes); err != nil {
-		return err
-	}
-
-	// Get script starting from the most recent OP_CODESEPARATOR.
-	subScript := vm.subScript()
-
-	// Generate the signature hash based on the signature hash type.
-	var hash []byte
-	if vm.isWitnessVersionActive(0) {
-		var sigHashes *TxSigHashes
-		if vm.hashCache != nil {
-			sigHashes = vm.hashCache
-		} else {
-			sigHashes = NewTxSigHashes(&vm.tx)
+	var sigVerifier signatureVerifier
+	switch {
+	// If no witness program is active, then we're verifying under the
+	// base consensus rules.
+	case vm.witnessProgram == nil:
+		sigVerifier, err = newBaseSigVerifier(
+			pkBytes, fullSigBytes, vm,
+		)
+		if err != nil {
+			vm.dstack.PushBool(false)
+			return nil
 		}
 
-		hash, err = calcWitnessSignatureHashRaw(subScript, sigHashes, hashType,
-			&vm.tx, vm.txIdx, vm.inputAmount)
+	// If the base segwit version is active, then we'll create the verifier
+	// that factors in those new consensus rules.
+	case vm.isWitnessVersionActive(BaseSegwitWitnessVersion):
+		sigVerifier, err = newBaseSigVerifier(
+			pkBytes, fullSigBytes, vm,
+		)
 		if err != nil {
-			return err
+			vm.dstack.PushBool(false)
+			return nil
 		}
-	} else {
-		// Remove the signature since there is no way for a signature
-		// to sign itself.
-		subScript = removeOpcodeByData(subScript, fullSigBytes)
-
-		hash = calcSignatureHash(subScript, hashType, &vm.tx, vm.txIdx)
 	}
 
-	pubKey, err := btcec.ParsePubKey(pkBytes, btcec.S256())
-	if err != nil {
-		vm.dstack.PushBool(false)
-		return nil
-	}
-
-	var signature *btcec.Signature
-	if vm.hasFlag(ScriptVerifyStrictEncoding) ||
-		vm.hasFlag(ScriptVerifyDERSignatures) {
-
-		signature, err = btcec.ParseDERSignature(sigBytes, btcec.S256())
-	} else {
-		signature, err = btcec.ParseSignature(sigBytes, btcec.S256())
-	}
-	if err != nil {
-		vm.dstack.PushBool(false)
-		return nil
-	}
-
-	var valid bool
-	if vm.sigCache != nil {
-		var sigHash chainhash.Hash
-		copy(sigHash[:], hash)
-
-		valid = vm.sigCache.Exists(sigHash, signature, pubKey)
-		if !valid && signature.Verify(hash, pubKey) {
-			vm.sigCache.Add(sigHash, signature, pubKey)
-			valid = true
-		}
-	} else {
-		valid = signature.Verify(hash, pubKey)
-	}
+	// TODO(roasbeef): verify NULLFAIL semantics as relates to constructors
+	// above and empty sig vectors
+	valid := sigVerifier.Verify()
 
-	if !valid && vm.hasFlag(ScriptVerifyNullFail) && len(sigBytes) > 0 {
+	switch {
+	case !valid && vm.hasFlag(ScriptVerifyNullFail) && len(fullSigBytes[1:]) > 0:
```

Comment:
>Do we slice the first byte off fullSigBytes because we know that we have at least one byte that corresponds to the hash type, so the remaining should be for the signature itself?

Yeah, before this PR, this fragment looked like: https://github.com/btcsuite/btcd/blob/81fbd9b67e5497963b089bd77288933e4f7657c6/txscript/opcode.go#L1966

At that point, `sigBytes` already has had the sighash byte removed from it. Open to ways to make this clearer or more explicit. It's one of the down stream effects of needing to explicitly deal with 64 vs 65 byte signatures unfortunately. 

> Interesting how a nil push (no sig, no hash type) with ScriptVerifyNullFail doesn't halt execution but this case does.

The null fail semantics are a bit tricky implementation wise. IMO the most straight forward statement is that: if a checksig is failing due to validation, then the signature must be an empty byte array. 

It comes up more in multi-sig settings where old scripts didn't have tapscript paths, so may be trying multiple multi-sig thresholds to see which one passes. W/o nullfail, some one could mutate sigs that weren't needed which then modifies the wtxid.